### PR TITLE
Create better error message for assignment on immutable - fixes #1110

### LINKED ISSUE #1110: Improve error reporting: usage of '<-' operator on non mutable bindings
## What

Sometimes assignment operator is used on immutable values, currently, for all cases the compilers gives `This value is not mutable` :

```fsharp
let x = 10
x <- 20
```

## Why

In cases where it is straightforward to devise a fix (changing simple value to mutable), giving help to developers unfamiliar with F# on how to change their code smooths the learning curve.

## How

The message could read `value 'x' is not mutable, please check that it is declared as mutable: 'let mutable x = 10' if you intend to mutate it.`

If it is not easily possible from AST to quote the RHS expression, we can still show the LHS.

### PR DESCRIPTION
**Pull request by forki**
image

Update: changed to wording proposed by @dsyme

### CODE REVIEW

File: src/fsharp/FSStrings.resx [hunk 1]
```diff
@@ -970,7 +970,7 @@
     <value>This rule will never be matched</value>
   </data>
   <data name="ValNotMutable" xml:space="preserve">
-    <value>This value is not mutable</value>
+    <value>Values are immutable by default. If you wish to mutate this value, you must explicitly declare it as a mutable variable using the mutable keyword, e.g. 'let mutable {0} = expression'.</value>
```
Comment (dsyme) [inline src/fsharp/FSStrings.resx]: One small modification please:

    Values are immutable by default. If you intend to to mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression'.

thanks

Comment (forki) [inline src/fsharp/FSStrings.resx]: done

File: src/fsharp/FSStrings.resx [hunk 2]
```diff
@@ -970,7 +970,7 @@
     <value>This rule will never be matched</value>
   </data>
   <data name="ValNotMutable" xml:space="preserve">
-    <value>This value is not mutable</value>
+    <value>Values are immutable by default. If you intend to mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression'.</value>
```
Comment (dsyme) [inline src/fsharp/FSStrings.resx]: Thanks for updating. Looking at this one last time, I wonder why not keep the original wording for the first part?  Somehow it feels better now I look at it yet again 

    This value is not mutable. If you intend to mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression'

c.f.

    Values are immutable by default. If you intend to mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression' 

Somehow the first describes the reality of the situation without needing to mention 'by default'?

Comment (forki) [inline src/fsharp/FSStrings.resx]: done

**Comment by dsyme** [comment]
OK, this looks good to go to me.

**Comment by dsyme** [comment]
One failure on CI part 2, I presume unrelated to this PR?

    fail001		CodeGen\EmittedIL\SteppingMatch	#  (SteppingMatch01.fs)

**Comment by forki** [comment]
Will take a look at this tomorrow.

**Comment by forki** [comment]
fixed and rebased. I think it's ready.

**Comment by KevinRansom** [comment]
The trouble with wordy stuff like error messages is that you end up bike-shedding them, here is my contribution.  I think the proposed message is better than the shipping message but is a touch informal for a compiler,

How about this:
````
Value is immutable, to enable mutation apply mutable to binding, e.g. 'let mutable v = 3'.
````

**Comment by smoothdeveloper** [comment]
@KevinRansom I agree that a terse (but explanative) message is better than too wordy. Is someone at Microsoft going to review all changed messages for consistency and style?

How about:

`Value is immutable, to mutate it, make the binding mutable, e.g. 'let mutable x = expression'`

I'm a bikeshedder too I guess :)

**Comment by forki** [comment]
There is no law that says a compiler needs formal error messages. AFAIK
that's more a historical thing ;-)

I think we should go similar route like Elm and concentrate on "correct +
helpful for newcomers".

And of course this is bikeshedding, but that's not too bad since it will
improve the situation eventually.
On May 5, 2016 4:34 AM, "Gauthier Segay" <notifications@github.com> wrote:

@KevinRansom <https://github.com/KevinRansom> I agree that a terse (but
explanative) message is better than too wordy. Is someone at Microsoft
going to review all changed messages for consistency and style?

How about:

Value is immutable, to mutate it, make the binding mutable, e.g. 'let
mutable x = expression'

I'm a bikeshedder too I guess :)

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/Microsoft/visualfsharp/pull/1147#issuecomment-217060941>

**Comment by dsyme** [comment]
The message in the PR is acceptable and consistent with the style of error messages already used in the compiler. (it's different to the screenshot above) I recommend we just go with that.

**Comment by KevinRansom** [comment]
I dislike the "if you intend to ' formulation as it is unnecessary and informal.   How about this it is exactly the same message as the PR which I otherwise liked without that bit?

This value is not mutable, to mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression'.</value> 

This bike shed is going to be mighty fine. :-)

**Comment by smoothdeveloper** [comment]
I find:
* the repetition close to one another of `this value` 
* not using `immutable` to avoid repeating four words starting with `muta`

is leading me to bike-shedding but otherwise the message in the PR is fine.

**Comment by smoothdeveloper** [comment]
Adopting new practice of compiler-warning-prose-golf.

**Comment by dsyme** [comment]
More bike sheds!

To be consistent with our other error messages, there needs to be a clear separation between the "error" and the "advice".  These should be in separate, well-formed sentences, so 

    This value is not mutable. To mutate this value, declare it using the mutable keyword, e.g. 'let mutable {0} = expression'.


not 

    This value is not mutable, to mutate this value...

In general the "advice" part of the error message should carry some text like "Consider ...." but it's not essential.

@smoothdeveloper Agreed the four "mut" words is not ideal but we can live with it here.

**Comment by KevinRansom** [comment]
Going with:
This value is not mutable. Consider using the mutable keyword, e.g. 'let mutable {0} = expression'

The bike shed will be:  'Burnt Sienna':  https://en.wikipedia.org/wiki/Sienna#Burnt_sienna_pigment_.28Maerz_and_Paul.29

Enjoy

**Comment by smoothdeveloper** [comment]
@dsyme thanks for explaining the guideline, @KevinRansom thanks for solving all my concerns with the message, I like "burnt sienna".

**Comment by isaacabraham** [comment]
Great to see another one of these going in :-) I do feel it's a shame that the "why" part of the error message has been omitted though i.e.the fact that all values are immutable by default in F#. There's still a good chance someone new to the language will see this error and understand what has happened, but not *why* it has happened.

**Comment by forki** [comment]
Yeah that's a bit unfortunate now. Sigh.

**Comment by dsyme** [comment]
@isaacabraham @forki I think you're referring to the "Values are immutable by default". I had asked for this to be removed in a line comment that's now lost from the record unfortunately.  My feeling was that it only raised more questions than it answered. Some explanation is crystal clear, but for that one you can just ask "why's it like that?" again, especially the "by default" part - which makes it sound like some arbitrary default in F# that could have been different.

I think the message is a clear improvement, and consistent with what we do today (error + advice), But I'll keep in mind that you're looking to explain "why" as well.  We should also look for existing error messages in F# that do explain "why" to use as examples - in general you are right that F# error messages often don't explain "why".

**Comment by dsyme** [comment]
To summarize, I think the "advice" section of the error message is a clear improvement in this PR: the error message is now actionable instead of leaving people scratching their heads.  

However we haven't added a "why" section.  I'm not intrinsically opposed to adding "why" sections, but they are hard to get right (and for this reason most of our error messages don't have them).
